external/mmat.m

```diff
@@ -51,9 +52,46 @@
 nDB = ndims(B);
 nD = max(nDA,nDB);
 
+if nD == 2 && isdefaultdim
+    C = A * B;
+    return
+end
+
 nA = [size(A),ones(1,nD-nDA)]; nA = nA(dim); 
 nB = [size(B),ones(1,nD-nDB)]; nB = nB(dim);
 
+neededMem = (numel(A)*nB(2) + numel(B)*nA(1)) * 16;
+if sw_freemem < neededMem && isdefaultdim
+    % Not enough memory to expand matrix to use bsxfun; use slow loop instead
+    szA = size(A); if numel(szA) > 3, A = reshape(A, [szA(1:2) prod(szA(3:end))]); end
+    szB = size(B); if numel(szB) > 3, B = reshape(B, [szB(1:2) prod(szB(3:end))]); end
+    if numel(szA) == 2 && numel(szB) > 2
+        output_shape = [szA(1) szB(2:end)];
+        C = zeros([szA(1) szB(2) size(B,3)]);
+        for ii = 1:size(B,3)
+            C(:,:,ii) = A * B(:,:,ii);
+        end
+    elseif numel(szA) > 2 && numel(szB) == 2
+        output_shape = [szA(1) szB(2) szA(3:end)];
+        C = zeros([szA(1) szB(2) size(A,3)]);
+        for ii = 1:size(A,3)
+            C(:,:,ii) = A(:,:,ii) * B;
+        end
+    else
+        output_shape = [szA(1) szB(2:end)];
+        if size(A, 3) ~= size(B, 3), error('Extra dimensions do not agree'); end
```

Comment:
I think this check could go on line 80 (i.e. `elseif size(A, 3) == size(B, 3)` - then throw an error on `else`) - does that make sense?